src/view-backend-exportable-private.h

```diff
@@ -54,7 +54,7 @@ class ClientBundle {
     uint32_t initialHeight;
 };
 
-class ViewBackend : public WS::ExportableClient, public FdoIPC::MessageReceiver {
+class ViewBackend final : public WS::ExportableClient, public FdoIPC::MessageReceiver {
```

Comment:
Oooh, multiple inheritance! WS::ExportableClient and FdoIPC::MessageReceiver still both need virtual destructors.

Reply:
Not really. A method (or destructor) needs to be marked `virtual` to make sure that the compiler ends up calling the one at the end of the inheritance chain when applying the `delete` operator to a pointer typed with one of the superclasses.

In the case of `ViewBackend`, the deletion is done in this way:

```cpp
auto* backend = reinterpret_cast<ViewBackend*>(data);
delete backend;
```

If the compiler knows that `ViewBackend` will not have subclasses (e.g. by marking ìt `final`) then it knows that it always needs to use `ViewBackend::~ViewBackend()` for destruction, and that this is always safe because there cannot be a more concrete implementation.

Reply:
Side note: If the destructor was marked `virtual` and at the same time the `ViewBackend` class is marked `final`, then the compiler would [devirtualize](https://marcofoco.com/the-power-of-devirtualization/) the call—which effectively ends up choosing which version of the method/destructor will be used statically at compile time (thus skipping the dynamic vtable lookup at runtime).

Reply:
> Here the compiler knows that we are deleting a `ViewBackend` (see the cast!), which will not ever have subclasses (when marked `final`). Therefore it knows that it always needs to use `ViewBackend::~ViewBackend()` for destruction, and that this is always safe because there cannot be a more concrete implementation.

I understand that, but if you ever change the way this ViewBackend instance is deleted, or if you ever create a different ViewBackend instance elsewhere and delete it from a parent class pointer, you leak the ViewBackend portion of the class, akin to slicing. This is much more of a footgun than leaving "final" off a struct declaration.

As a rule, any class that is derived from should have a virtual destructor. This is accepted C++ best-practice for decades now and is also the rule we use in WebKit.

Reply:
There's only two ways in the way that the `ViewBackend` can be destroyed:

- A subclass is added, deletion still done through a `ViewBackend*` pointer: the compiler will tell you to remove final, and then warn about the class needing a virtual destructor. No problem.
- No subclass is added, but the destruction is done through a superclass pointer (`WS::ExportableClient*` or `FdoIPC::MessageReceiver`): the compiler will tell you that you need a virtual destructor because the type is not a final class. Again: no problem.

So there's no footgun now that we have compiler warnings enabled. How is the so-called “accepted best practice” not just cargo-culting in this case?

P.S: There is not such thing as “leaking” a slice of an object itself as a result from the wrong constructor being chosen, what happens is that some constructors are just not executed—if the code that does not run happens to dispose of resources those won't be properly disposed (of course: not cool), but the whole zone of memory allocated for *the instance itself* is always freed in full.